### ISSUE TITLE
Variables declared in a file that is conditionally imported via a mixin cannot be overridden in the global scope. Any alternatives?

### ISSUE THREAD

**Issue by folosophy** [less/less.js#3080]
If I import a file using a mixin like so...

```
.Nav-Main();
@Nav-Main: on;

.Nav-Main () when (@Nav-Main = on) {
  @import 'nav-main.less';
}
```

...and then set a variable in that file, that variable is NOT part of the global scope because it's now scoped within Nav-Main. I can't override any of the variables I declare in that file.

So... is there any way to conditionally import a file, but still be able to override the variables you declare inside it in the global scope?

**Comment by seven-phases-max** [comment]
There are alternatives, but it's not clear why to use import or not import depending on variable on/off instead of lack or precence of tehe import itself. Describe your use case - otherwise it looks like an XY-problem.

**Comment by folosophy** [comment]
Ok, so I've built a modular library with lots of different options for what to import or not import.

For example, if the variable "nav-main-type" is set to "mobile", then the LESS file for the mobile menu gets imported normally, but if it's set to "dual", then it gets imported within a "max-width: 1200px" media query and the desktop menu gets imported inside a "min-width: 1201px" query.

If I have a site that needs a highly custom nav menu, I set "Nav-Main" to "off", which doesn't import "nav-main" at all.

Basically, I'd like to be able to turn large portions of LESS code and/or files on and off without running into scope issues.

**Comment by seven-phases-max** [comment]
I see (well, I also have already seen https://stackoverflow.com/questions/44529344 so I've got a basic idea of what you're trying to do). So how do you change these `nav-main-type`/`Nav-Main`/etc. values for different builds? (It's not that this knowledge would drastically change something - but I'd like to know as much as possible because my answer is going to be a waaaaaay very long to explain why such variable-based-conditional-imports-based-configuration would always be a bad idea in Less, thus having to suggest some food/ideas for thought of rethinking the whole approach from scratch).

**Comment by folosophy** [comment]
So I have one folder for my LESS library, then another for whatever theme-specific LESS I need.

I import the main library file and then import my theme-specific LESS afterwards, so I can override any variables I need to.

If I'm doing a completely custom nav for a certain project, then I just set "@Nav-Main: off" and the library doesn't import any of the nav-main stuff.

I work around the scoping issue by putting all my variables that might need to be overridden in one of the main library files, like nav.less.

While this isn't 100% ideal (I'd rather define them in the sub-files), I've used the conditional importing technique for my past 6 projects and it's worked out great.

Definitely interested to hear counter-arguments and/or other ideas, though.

**Comment by rjgotten** [comment]
> other ideas

- Place the Less for your main nav inside the imported library file into mixins.
- Import the library file statically, at the top outside of any selectors or mixins.
- Call the mixins imported from the libary where you are currently doing the dynamic import.

You can then either continue to overwrite global variables -- because now the file _is_ imported in the global scope -- or you can do things the proper way and pass arguments to the mixins to control what gets written out.

**Comment by matthew-dean** [comment]
What I would suggest doing is having a "start" mixin that loads the remainder of selectors / variables. I've used a similar approach in the past.

As in:
```less
.Nav-Main();
@Nav-Main: on;

.Nav-Main () when (@Nav-Main = on) {
  @import 'nav-main.less';
  .init();
}
.Nav-Main() when (default()) {
  .init();
}
.init() {
  // do everything
}
```
That way, everything is evaluated in the same scope.
